### ISSUE TITLE
Problem with `RefreshScope`, centralized configuration and type safe configuration

### ISSUE THREAD

**Issue by bleporini** [spring-cloud/spring-cloud-commons#328]
Spring Boot 2.0.0.RC1 - Spring Cloud Finchley.M6 - JDK8

Hello,

Consider 2 SpringBoot applications (config server + app) as defined in https://spring.io/guides/gs/centralized-configuration/

When you map configuration to a bean with `@ConfigurationProperties` and you try to change it runtime here is the behaviour I found:

First time you update the property in the config server it works
Second time you update the property it works for properties injected with `@Value` but it doesn't work for those injected as  `@ConfigurationProperties`. 

You can find a sample project in https://github.com/bleporini/refresh .

```java
@RestController
@RefreshScope
@Slf4j
public class MyController {

    private final Features features;
    private final String test;

    public MyController(Features features,
                        @Value("${features.test}") String test) {
        this.features = features;
        this.test = test;
    }

    @Scheduled(fixedDelay = 1000)
    public void trace(){
        log.error("{} / {}", features.getTest(), test);
    }

    @GetMapping("/o")
    public String hello(){
        return features.getTest();
    }

    @ConfigurationProperties("features")
    @Slf4j
    public static class Features {

        @Getter
        private String test;

        public void setTest(String test) {
            log.warn("{} Was/ will be : {} / {}", identityHashCode(this), this.test, test);
            this.test = test;
        }
    }

}
```

It starts with a value set to `Hello1`:

```
2018-02-20 18:45:18.056 ERROR 58210 --- [pool-2-thread-1] com.example.refresh.MyController         : Hello1 / Hello1
```

Fine

Now we set the value to `Hello12` and it still works fine, except now the bean is lazily initialized (that's why there's a `@GetMapping` method in it but it is another story), so after waking up the bean with a curl:

```
2018-02-20 18:45:46.358 ERROR 58210 --- [pool-2-thread-1] com.example.refresh.MyController         : Hello12 / Hello12
```

Second time we update back the value to `Hello1`, the `@Value` is updated but not the bean, but now the bean is not lazily initialized, so no need to curl the app :

```
2018-02-20 18:45:58.394 ERROR 58210 --- [pool-2-thread-1] com.example.refresh.MyController         : Hello12 / Hello1
```

About the lazy initialisation after the first update, I think this is also an issue because in the sample there's a `@Scheduled` method, so after the first update, the bean seems to be not bound in the application context, and thus the scheduled method is not called till the bean is bound.

Note that the problem with the configuration properties disappears if you downgrade to Spring Boot 1.5.10 and Spring Cloud Edgware.SR2.

Note that the problem with the laziness is still the same with Spring Boot 1.5.10 and Spring Cloud Edgware.SR2. 

Regards

**Comment by ryanjbaxter** [comment]
Does using Consul matter or not?  Does the same problem occur regardless of the environment repo?

**Comment by bleporini** [comment]
At first I tried with Consul and it was the same. I tried then with Config server. The only difference is that with config server, I have to invoke actuator refresh endpoint, and with Consul it's not useful as it comes with a config watcher.

**Comment by spencergibb** [comment]
I think it has to do with proxies and your projects odd layout. Moving Features to top level seemed to fix things for me.

**Comment by thundur** [comment]
I'm experiencing similar behavior after upgrading my project to Spring Boot 2 and Spring Cloud Finchley.M7. I also have an example project: https://github.com/thundur/demo. Be sure to change the spring.config.additional-location property in the bundled bootstrap.properties file before running the app.

The ```@Value``` annotated property refreshes all the time, but the ```@ConfigurationProperties``` annotated class refreshes only the first time.

I compared debug logging between the first refresh and the subsequent refreshes, and during the first refresh, these lines are logged (you can find the logs attached):
```
[...]
DEBUG     o.s.b.f.s.DefaultListableBeanFactory - Returning cached instance of singleton bean 'configurationPropertiesRebinder'
DEBUG     o.s.b.f.s.DefaultListableBeanFactory - Returning cached instance of singleton bean 'loggingRebinder'
[...]
DEBUG     o.s.b.f.s.DefaultListableBeanFactory - Returning cached instance of singleton bean 'configurationPropertiesRebinder'
```
Subsequent refreshes don't log these lines. All other logging is the same (except for some class ids).

[firstRefresh.txt](https://github.com/spring-cloud/spring-cloud-commons/files/1774499/firstRefresh.txt)
[secondRefresh.txt](https://github.com/spring-cloud/spring-cloud-commons/files/1774501/secondRefresh.txt)

When downgrading to Boot 1.5.11 and Cloud Edgware.SR2, the refresh works properly.

**Comment by thundur** [comment]
Looks like the same issue is also reported in #290

**Comment by bleporini** [comment]
@spencergibb Can you please give clues about the odd layout? You say that because `Features` is an inner class?

**Comment by spencergibb** [comment]
@bleporini it was an inner class, yes. Moving it to top level fixed it for me.

**Comment by sanderino666** [comment]
@spencergibb in the example provided by @thundur this does not seem to be the case. Any idea what this problem could be?

**Comment by spencergibb** [comment]
I haven't been able to follow up yet, so I don't know

**Comment by bleporini** [comment]
@spencergibb I tried to move the configuration properties `Features` class at a upper level, but it's still the same: the configuration update is propagated to the `Features` class only after the first update, after next updates it doesn't change.

Did you change something else?

**Comment by m-van-tilburg** [comment]
Have the same issue with Spring Boot 2.0.0.RELEASE and Spring Cloud Finchley.M8. This is really blocking for upgrading my Spring Boot 1.5 based application to Spring Boot 2.0 :(

**Comment by spencergibb** [comment]
@thundur I can't make heads or tails of your demo. You have config server properties, but not a config server.

**Comment by spencergibb** [comment]
@m-van-tilburg looking for an application that reproduces the problem still.

**Comment by spencergibb** [comment]
@bleporini I upgraded to boot rc2 and finchley m7. Just updated to boot RELEASE and Finchley M8. works fine.

**Comment by sanderino666** [comment]
@spencergibb I am debugging the test setup from @thundur and there is indeed a problem. When running the application and executing the /refresh endpoint, the instance id of ConfigurationPropertiesRebinder which is created in the ConfigurationPropertiesRebinderAutoConfiguration is not the same as the instance id when adding a breakpoint in ConfigurationPropertiesRebinder.rebind(). In the rebind() it keeps referring to the first created instance and not the one which is initialized after refresh. In the rebind() the old applicationContext is being used which in turn holds old state.

I am debugging a bit further, but if you have some insights on my findings, please share :)

**Comment by spencergibb** [comment]
@sanderino666 the app isn't valid in my mind.

**Comment by sanderino666** [comment]
@spencergibb It could indeed be the case. The boot 1.5.x setup worked (could be by accident though :)) so I am looking for the missing piece of the puzzle. You mentioned _You have config server properties, but not a config server._. What do you mean by this?

**Comment by thundur** [comment]
@spencergibb, the app uses an embedded config server without endpoints, as mentioned in the last paragraph of https://cloud.spring.io/spring-cloud-config/multi/multi__embedding_the_config_server.html

**Comment by spencergibb** [comment]
@thundur those docs say an app should have `@EnableConfigServer` and yours doesn't, hence my confusion.

**Comment by thundur** [comment]
The last sentence of the page I posted states that `@EnableConfigServer` is not required imo.
And just like @sanderino666 states, the app works with spring boot 1.5.x, but not with spring boot 2.

The setup enables me to change the config on disk and use the refresh endpoint to reload the configuration.

**Comment by spencergibb** [comment]
ok

**Comment by spencergibb** [comment]
@thundur I'm reproducing #290

**Comment by spencergibb** [comment]
interestingly, the value is correct in `/actuator/env`, so the spring environment is getting updated, just not the config props bean.

**Comment by sanderino666** [comment]
@spencergibb I know, weird right? It feels like there are 2 application contexts present after a refresh. Could that be the case?

**Comment by spencergibb** [comment]
there are always two, bootstrap and the main, but the appropriate application context is being used and its environment is correct.

**Comment by sanderino666** [comment]
@spencergibb but when the rebind() fires, it is using a different application context then the one that is being used in ConfigurationPropertiesRebinderAutoConfiguration.afterSingletonsInstantiated() so it seems like there is more than 1 ConfigurationPropertiesRebinder which is weird because it is a singleton. The only explanation for that is that it resides in both contexts but is not being refreshed in both contexts.

**Comment by thundur** [comment]
@spencergibb I also saw that behavior. The refresh detects the property change, but the `@ConfigurationProperties` classes do not reflect the changed property from the second refresh on.
Glad you were able to reproduce the issue for #290, as I think it is really similar to mine.

**Comment by spencergibb** [comment]
See https://github.com/spring-projects/spring-boot/issues/12451

**Comment by spencergibb** [comment]
This issue will be fixed by [Spring Boot 2.0.1](https://github.com/spring-projects/spring-boot/milestone/97) due Mar 29. That should be what Finchley.RC1 builds upon.

**Comment by bleporini** [comment]
@spencergibb I'll check with Spring Boot 2.0.1 after Mar 29, but in the mean time, FYI I checked with 2.0.0.RELEASE and M8 and I did not see any improvement:
```
$ curl  localhost:8080/o
Hello1%
```
Updating `features.test` key in the config server repository and refreshing application context:
```                                                                                                                                                                                                                                                                          $ curl -XPOST localhost:8080/actuator/refresh -ksv
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> POST /actuator/refresh HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200
< Content-Type: application/vnd.spring-boot.actuator.v2+json;charset=UTF-8
< Transfer-Encoding: chunked
< Date: Mon, 26 Mar 2018 13:22:58 GMT
<
* Connection #0 to host localhost left intact
["features.test"]%                                                                                                                                                                                                                                                              $ curl  localhost:8080/o
Hello1%
$
```
After restarting the application:
```
$ curl  localhost:8080/o
Hello12%
```
That confirm the value has correctly been updated in  the repository and exposed by the config server.

**Comment by spencergibb** [comment]
There won't be any improvement until boot 2.0.1

**Comment by sanderino666** [comment]
@bleporini @spencergibb I tested the demo project from @thundur with Spring Boot 2.0.1.BUILD-SNAPSHOT and this works as expected.

**Comment by thundur** [comment]
@spencergibb, the issue is solved for me, thanks!

**Comment by forTechGig** [comment]
im using Finchley.M9 and 2.0.1.RELEASE, it did refresh when i tested in lower cycle.. i changed yml and tried refresh multiple times and those worked.. 
Now in production, few times it was okay but now one instance is getting refreshed and one instance is not.. but when i try manage/refresh, it gives me the attribute that has been updated.. anybody has this behaviour..

**Comment by spencergibb** [comment]
Finchley.RELEASE with boot 2.0.3 is the latest. It was a boot issue

**Comment by forTechGig** [comment]
thanks.. will test and update here..

**Comment by erindas** [comment]
Any updates if the fix worked for anybody? I am currently working on upgrading our spring boot application to version 2.0.6 and Finchley.RELEASE. But the auto-refresh feature of config server still doesn't work.

**Comment by ryanjbaxter** [comment]
Finchley.SR2 is the latest release.  Try that and if it doesnt work please open another issue with a sample that reproduces the problem.

**Comment by sguntaka** [comment]
Is this issue resolved? In fact I am facing the same issue where "/refresh" endpoint properties are getting updated in environment and showing updated values in "/actuator/env", but not in config props bean. I tried with **Spring Boot 2.1.6.RELEASE** and **Greenwich.SR2**. I have reverted spring boot to 1.5.22.RELEASE and spring cloud to Edgware.SR2. I have opened an issue with sample code. Any updates to this fix is really appreciated. Thank you.
[https://github.com/spring-cloud/spring-cloud-config/issues/1450#](url)

**Comment by sguntaka** [comment]
Can someone share Spring Boot and Spring Cloud versions that are worked for this fix?. I tried Spring boot version 2.0.6 and Spring Cloud Finchley.SR2 is not working. Any updates to this fix is much appreciated.

**Comment by erindas** [comment]
@sguntaka, try setting `spring.cloud.bus.id` in bootstrap or application properties. That worked for me.

**Comment by ryanjbaxter** [comment]
@sguntaka yes this was fixed, if you are still having a problem please open a separate issue.

**Comment by ankitkatiyar25** [comment]
@ryanjbaxter : Which version has the fix?

**Comment by ryanjbaxter** [comment]
It was fixes in 2.0.0.RELEASE that correlates to Finchley.RELEASE
